src/System.IO.Pipelines/Pipe.cs

```diff
@@ -393,6 +432,9 @@ public void AdvanceReader(ReadCursor consumed, ReadCursor examined)
 
             if (!consumed.IsDefault)
             {
+                var consumedBytes = ReadCursor.GetLength(_readHead, _readHead.Start, consumed.Segment, consumed.Index);
+                Interlocked.Add(ref _length, -consumedBytes);
```

Comment:
The locking around length is inconsistent. In some cases you update it inside of _sync and in this case you're using an interlocked. Move this code lower into the lock.